plonky2x/src/frontend/eth/mpt/mpt.rs

```diff
@@ -0,0 +1,310 @@
+            storage_proof.len(),
+            PROOF_LEN
+        );
+    }
+
+    let mut padded_elements = vec![vec![0u8; ENCODING_LEN]; PROOF_LEN];
+    let mut lengths = vec![0usize; PROOF_LEN];
+
+    for (i, inner_vec) in storage_proof.into_iter().enumerate() {
+        // Check inner length
+        if inner_vec.len() > ENCODING_LEN {
+            panic!(
+                "Proof element {} is too long, has {} elements, but ENCODING_LEN is {}",
+                i,
+                inner_vec.len(),
+                ENCODING_LEN
+            );
+        }
+        lengths[i] = inner_vec.len();
+        for (j, &byte) in inner_vec.iter().enumerate() {
+            padded_elements[i][j] = byte;
+        }
+    }
+
+    (padded_elements, lengths)
+}
+
+impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
+    pub fn le(&mut self, lhs: Variable, rhs: Variable) -> BoolVariable {
+        let generator: LeGenerator<L, D> = LeGenerator {
+            lhs: lhs.clone(),
+            rhs: rhs.clone(),
+            output: self.init::<BoolVariable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    pub fn byte_to_variable(&mut self, lhs: ByteVariable) -> Variable {
+        let generator: ByteToVariableGenerator<L, D> = ByteToVariableGenerator {
+            lhs: lhs.clone(),
+            output: self.init::<Variable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    pub fn sub_byte(&mut self, lhs: ByteVariable, rhs: ByteVariable) -> ByteVariable {
+        let generator: ByteSubGenerator<L, D> = ByteSubGenerator {
+            lhs: lhs.clone(),
+            rhs: rhs.clone(),
+            output: self.init::<ByteVariable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    #[allow(dead_code, unused_variables)]
+    pub fn to_nibbles(&mut self, bytes: &[ByteVariable]) -> Vec<ByteVariable> {
+        let len = bytes.len() * 2;
+        let generator: NibbleGenerator<L, D> = NibbleGenerator {
+            input: bytes.to_vec(),
+            output: (0..len).map(|_| self.init::<ByteVariable>()).collect_vec(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    const PREFIX_EXTENSION_EVEN: u8 = 0;
+    const PREFIX_EXTENSION_ODD: u8 = 1;
+    const PREFIX_LEAF_EVEN: u8 = 2;
+    const PREFIX_LEAF_ODD: u8 = 3;
+    /// Get the validators for a given block root.
+    /// P is the number of proof elements to be considered
+    pub fn verify_mpt_proof<const ENCODING_LEN: usize, const PROOF_LEN: usize>(
+        &mut self,
+        key: Bytes32Variable,
+        proof: ArrayVariable<ArrayVariable<ByteVariable, ENCODING_LEN>, PROOF_LEN>,
```

Comment:
Instead of ArrayVariable<ByteVariable, ENCODING_LEN>

use BytesVariable<ENCODING_LEN>?

Reply:
I think `BytesVariable` should just be an `ArrayVariable` of `Byte` actually so that the indexing ops from `ArrayVariable` can flow through and mux also works on it. Is there a nice way to do type alias in Rust?

Reply:
Type alias seems unecessary, just array directly?